packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -0,0 +1,395 @@
+
+/// A table that shows [TreeNode]s.
+///
+/// This takes in a collection of [data], and a collection of [ColumnData].
+/// It takes at most one [TreeColumnData].
+///
+/// The [ColumnData] gives this table information about how to size its
+/// columns, how to present each row of `data`, and which rows to show.
+///
+/// To lay the contents out, this table's [_TreeTableState] creates a flattened
+/// list of the tree hierarchy. It then adds the nesting level of the
+/// deepest row in [data] that is currently visible with the prescribed
+/// width of each row in the table.
+class TreeTable<T extends TreeNode<T>> extends StatefulWidget {
+  const TreeTable({
+    Key key,
+    @required this.columns,
+    this.treeColumn,
+    @required this.data,
+    @required this.keyFactory,
+  }) : super(key: key);
+
+  /// The columns to show in this table.
+  final List<ColumnData<T>> columns;
+
+  /// The column of the table to treat as expandable.
+  final TreeColumnData<T> treeColumn;
+
+  /// The tree structure of rows to show in this table.
+  final T data;
+
+  /// Factory that creates keys for each row in this table.
+  ///
+  /// It uses [PageStorageKey] to restore the table's state when re-opening
+  /// its page.
+  final String Function(T data) keyFactory;
+
+  @override
+  _TreeTableState<T> createState() => _TreeTableState<T>();
+}
+
+class _TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
+    with TickerProviderStateMixin {
+  List<T> _flattenedList = [];
+
+  /// The width of each column in this table, including the
+  /// appropriate indentation for nested rows.
+  ///
+  /// This is computed in [_computeTableWidth].
+  List<double> columnWidths = [];
+  double get tableWidth => columnWidths.reduce((x, y) => x + y);
+
+  AnimationController _resizeController;
+  Animation<double> _resizeAnimation;
+  Tween<double> _widthTween;
+
+  /// The width to assume for columns that don't specify a width.
+  static const defaultColumnWidth = 500.0;
+
+  /// The maximum height to allow for rows in the table.
+  static const defaultRowHeight = 42.0;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.data.expandCascading();
+    _flattenedList = _flattenExpandedTree();
+    columnWidths = _computeTableWidth();
+    _widthTween = Tween<double>(begin: tableWidth, end: tableWidth);
+    _resizeController = AnimationController(
+      vsync: this,
+      duration: const Duration(milliseconds: 200),
+    );
+    _resizeAnimation = CurvedAnimation(
+      parent: _resizeController,
+      curve: Curves.easeInOutCubic,
+    );
+  }
+
+  @override
+  void didUpdateWidget(TreeTable<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _flattenedList = _flattenExpandedTree();
+    refreshTree();
+  }
+
+  @override
+  void dispose() {
+    _resizeController.dispose();
+    super.dispose();
+  }
+
+  List<T> _flattenExpandedTree({T node, bool Function(T node) filter}) {
+    node ??= widget.data;
```

Comment:
This is an O(n^2) algorithm for getting the flattened children given deep trees. The TreeNode class used here has a convenience hooks to do this more efficiently.  You can do it in O(n) time and perhaps less time not that it really matters by being clever about computing indexes. InspectorTree shows some examples of the technique.

Reply:
Can you comment `done` on all resolved comments to make it clear which ones are pending?

Reply:
I clearly didn't have enough coffee yesterday if I'm writing that expensive of a tree traversal 😢